snappy/snapp.go

```diff
@@ -1272,34 +1224,33 @@ func (s *SnapPart) CanInstall(allowOEM bool, inter interacter) error {
 	return nil
 }
 
-var timestampUpdater = helpers.UpdateTimestamp
-
-func updateAppArmorJSONTimestamp(fullName, thing, version string) error {
-	fn := filepath.Join(dirs.SnapAppArmorDir, fmt.Sprintf("%s_%s_%s.json", fullName, thing, version))
-	return timestampUpdater(fn)
-}
-
-// RequestAppArmorUpdate checks whether changes to the given policies and
-// templates impacts the snap, and updates the timestamp of the relevant json
-// symlinks (thus requesting aaClickHookCmd regenerate the appropriate bits).
-func (s *SnapPart) RequestAppArmorUpdate(policies, templates map[string]bool) error {
-
-	fullName := QualifiedName(s)
+// RequestSecurityPolicyUpdate checks whether changes to the given policies and
+// templates impacts the snap, and updates the policy if needed
+func (s *SnapPart) RequestSecurityPolicyUpdate(policies, templates map[string]bool) error {
+	var foundError error
 	for _, svc := range s.ServiceYamls() {
 		if svc.NeedsAppArmorUpdate(policies, templates) {
-			if err := updateAppArmorJSONTimestamp(fullName, svc.Name, s.Version()); err != nil {
-				return err
+			err := svc.generatePolicyForServiceBinary(s.m, svc.Name, s.basedir)
+			if err != nil {
+				logger.Noticef("Failed to regenerate policy for %s: %v", svc.Name, err)
+				foundError = err
```

Comment:
A bit strange that it continues but only the last error is tracked. Makes it look like the problem was smaller than it actually was. One option in these cases is to build an error value which is an aggregation of all found errors. Please ping me for a real example doing that if you're interested.